### PR TITLE
Initial support of Gradle's model for C++

### PR DESCRIPTION
**Pull request by Shad0w1nk**
This is the first step toward supporting the new model in Gradle. More works need to be done to better integrate the code with the Java plugin, however it was starting to be a bit intrusive and I wanted to cut a release and keep working on the detail later. This pull request contains the following:
- Upgrade the wrapper to the latest release, Gradle 2.7
- tools in ProtobufConfigurator was changed to protected as it should not be available to the public but seems to be needed in the model. How the extension is registered differ from Java which make the statement protobuf.tools.resolve() fail in the model as the member variable is private. A better way for fixing this problem should be implemented.
- The ProtobufPlugin class was modified to choose between the Java/Android or model (c++) implementation. As much code as possible was shared between both world meaning we are reusing the task and the convention. Unfortunately, the source set use a separate class in order to support the model.
- Integration test were added using TestKit available in the latest Gradle. Some wiring of tasks had to be done according to the documentation.
- A submodule to a release version of Protobuf was added to make the C++ work without having to do any kung fun with installing library. Gradle will build protoc and protobuf library from source and use them in the example project and Gradle TestKit. All the wiring of the building has been done so executing "./gradlew test" will automatically build the necessary protobuf binary. As long as gcc (linux), clang (mac os) and msvc 2013 (windows) is installed, everything will work out of the box. A check has been added if someone forgot to sync the submodule.

What's the next step? The logic for building protobuf is a bit different then the legacy Java as the model does add some flexibility that needs to be taken into account (see the integration test for example of what can be done). Some discussion and more work will need to be done in order to prepare the Java plugin for the model and integrate more code with that plugin.

Any input is welcome and I will fix anything that need to be fixed.

### CODE REVIEW

**Comment by wrwg** [comment]
Hi,
is this PR still active? I'm rather interested in making this plugin Java neutral. Some thoughts:

- Can this be done so we don't need to hard wire the different languages into the same project? I guess that's the intention anyway, but just to be sure.
- I feel it would be cleaner we decouple the plugin from protoc, i.e. let protoc just generate the descriptor and call the plugin as a standalone program. This also would better support other tools working on the descriptor. @pherl, is this technically possible? 

@zhangkun83 @ejona86 @louiscryan @pherl

**Comment by Shad0w1nk** [comment]
I got the feeling from Kun the project was open to support C++ (https://github.com/google/protobuf-gradle-plugin/issues/21). The new C++ model in Gradle will become the standard for the Java project too. Most of the work for C++ is beneficial for Java.

This PR is making the plugin friendly with the native ecosystem already in Gradle. If you only apply the Java plugin then only the Java DSL and task will be wired in your project. If you decide to go with C++ then everything will work out of the box just as it does with Java.

This PR is the first step in making the plugin work with several language. Some polishing needs to be done to minimize the code duplication between the two languages and make other language easier to incorporate.

The way Gradle handle tasks, it's easy to hook post processing task that could be require for working on the descriptor.

**Comment by zhangkun83** [comment]
Sorry for not responding. There is a problem in my inbox filtering so I didn't see this PR. I will take a look ASAP.

**Comment by wrwg** [comment]
Thanks Kun for looking into Daniel's PR.

As regards the descriptor thing. I also need to run protoc with just
outputting the descriptor and not generating code. I think a much cleaner
design than the current plugin model of protoc would be a pipeline model,
as most compilers do, and which integrates well with any build system.
Anyway, I can live with the slight awkwardness of working around this.

**Comment by zhangkun83** [comment]
@wrwg in case you didn't know, you can disable all the codegen outputs and use `task.generateDescriptorSet = true` to output just the descriptors.

File: build.gradle [hunk 1]
```diff
@@ -9,6 +12,11 @@ Base directory: $projectDir
 Running script ${relativePath(buildFile)}
 """
 
+if (file("vendor/protobuf").list().length == 0) {
+	throw new GradleException(
```
Comment (zhangkun83) [inline build.gradle]: code style: please use 2-space indent instead of tab.

**Comment by zhangkun83** [comment]
I tried to test your branch. It fails with:
```
com.google.protobuf.gradle.ProtobufModelPluginTest > a default protobuf source set exists for configuration FAILED
    org.spockframework.runtime.ConditionNotSatisfiedError at ProtobufModelPluginTest.groovy:42
...
```
Is the protoc binary supposed to be built if I run `gradle test`?

File: src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+
+  def "a default protobuf source set exists for configuration"() {
+    given: "a native executable with proto source set configuration"
+    buildFile << """
+          model {
+            components {
+              main(NativeExecutableSpec) {
+                sources {
+                  proto {
+                    source.include "*.proto"
```
Comment (zhangkun83) [inline src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy]: Can we include `*.proto` by default so you don't even need the `proto {}` block in the default case?

File: build.gradle (hunk 1 above)
Comment (Shad0w1nk) [inline build.gradle]: I followed the convention used in this file, do you want me to fix the indentation inside the entire file?

File: src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy (hunk 2 above)
Comment (Shad0w1nk) [inline src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy]: The include `*.proto` is actually not needed here. I think I wanted to be implicit. The source set follow the same pattern as every other source set in Gradle meaning by default it will include any file under the language directory. The `*.proto` is a configuration taste for the user. I will remove where it's not needed so it doesn't confuse anyone.

Comment (Shad0w1nk) [inline src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy]: As for the failure during `gradle test`, make sure you use the wrapper (`gradlew test`). Also could you tel me on what machine you tried. I did `gradlew test` on the following combination and they all worked:
 * Windows 10 with msvc 2013
 * Ubuntu 14.04 with gcc 4
 * Mac OS X 10.10 with clang 7

File: build.gradle (hunk 1 above)
Comment (Shad0w1nk) [inline build.gradle]: If you have nothing against it, I can incorporate the codenarc plugin with the specification you told me for the style. This way we won't have the styling problem anymore as Gradle will catch them. What do you think?

**Comment by Shad0w1nk** [comment]
I understand what you mean @wrwg. The way I see the philosophy behind Gradle, when applying the plugin protobuf, it could add a preconfigured task just for generating the descriptor set. This way any post processing that needs to be done could simply depends on the that task and wire it inside your workflow. With the legacy java, this additional task would have some overhead as the tasks are eagerly configured. However with the new model, the task are lazily configured which add no overhead during configuration if it's not referenced. The recommendation of @zhangkun83 does work too.

File: build.gradle (hunk 1 above)
Comment (zhangkun83) [inline build.gradle]: That sounds good, but it that would update existing code too, we may want to do it a separate commit. We don't want to mix reformatting and real change in the same commit.

File: src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy (hunk 2 above)
Comment (zhangkun83) [inline src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy]: Yes I am running the wrapper.

If I run it under the root project directory, it won't compile protoc. If I test it under `testProjectNative`, it will try to build protoc, but still end up with linkage errors:
```
linking protoc failed.
/home/zhangkun/github/protobuf-gradle-plugin-shad0w1nk/vendor/build/binaries/protocLibStaticLibrary/libprotocLib.a(command_line_interface.o): In function `google::protobuf::compiler::CommandLineInterface::MemoryOutputStream::MemoryOutputStream(google::protobuf::compiler::CommandLineInterface::GeneratorContextImpl*, std::string const&, bool)':
command_line_interface.cc:(.text+0x1383): undefined reference to `google::protobuf::io::StringOutputStream::StringOutputStream(std::string*)'
command_line_interface.cc:(.text+0x13f6): undefined reference to `google::protobuf::io::ZeroCopyOutputStream::~ZeroCopyOutputStream()'
/home/zhangkun/github/protobuf-gradle-plugin-shad0w1nk/vendor/build/binaries/protocLibStaticLibrary/libprotocLib.a(command_line_interface.o): In function `google::protobuf::compiler::CommandLineInterface::MemoryOutputStream::MemoryOutputStream(google::protobuf::compiler::CommandLineInterface::GeneratorContextImpl*, std::string const&, std::string const&)':
command_line_interface.cc:(.text+0x14a6): undefined reference to `google::protobuf::io::StringOutputStream::StringOutputStream(std::string*)'
command_line_interface.cc:(.text+0x1519): undefined reference to `google::protobuf::io::ZeroCopyOutputStream::~ZeroCopyOutputStream()'
```

File: build.gradle (hunk 1 above)
Comment (Shad0w1nk) [inline build.gradle]: Sounds good, I will fix the line space and look at the coding style in another PR.

File: src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy (hunk 2 above)
Comment (Shad0w1nk) [inline src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy]: I confirm the two problems you just reported... :-( I will look into fixing them as soon as possible.

**Comment by Shad0w1nk** [comment]
I finally had time to test on all major OS (Windows, Mac OS X and Linux). All test pass and should now work perfectly. If you have anything left to fix, tell me so I can fix them.

**Comment by zhangkun83** [comment]
Just wanted you to know that I am not dropping this. I have been very busy recently. Hopefully I will have time to look at it by this weekend.

**Comment by Shad0w1nk** [comment]
No worries, I can relate.

File: build.gradle [hunk 3]
```diff
@@ -61,16 +69,46 @@ repositories {
...
+
+  doLast {
+    outputDir.mkdirs()
+    file("$outputDir/plugin-classpath.txt").text = sourceSets.main.runtimeClasspath.join("\n")
+  }
+}
+
+tasks.test.dependsOn tasks.createClasspathManifest
+if (findProject(':vendor') != null) {
+  def vendorModelRegistry = project.evaluationDependsOn(':vendor').getModelRegistry()
```
Comment (zhangkun83) [inline build.gradle]: `evalutionDependsOn()` returns `Project`, but I cannot find `getModelRegistry()` on `Project`. What's happening here?

Comment (zhangkun83) [inline build.gradle]: Oh, it's on `ProjectInternal`. `ModelRegistry` is also internal. Can you give me more background on what you are doing in this if-statement and why?

Comment (Shad0w1nk) [inline build.gradle]: Since I'm using the new TestKit from Gradle and I require the protoc and protobuf library in the test, I need them to be built before I run the test. This if-statement ensure these requirements are meet for the test task which execute the TestKit test. I force the model to "realize" (aka instantiate) the required node from the project 'vendor'. I can then get the binaries instances and add them as a dependency of the test task. When the test task is executed, the protoc and protobuf library will get build before.

I'm aware the ModelRegistry is a internal feature and I'm actively searching for an alternative in the public API. I'm hoping to find a better solution in my next PR.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+    // Provides the osdetector extension
+    project.pluginManager.apply('osdetector')
+
+    project.extensions.create("protobuf", ProtobufConfigurator, project, fileResolver);
+
+    project.pluginManager.apply(ComponentModelBasePlugin)
+  }
+
+  static class Rules extends RuleSource {
```
Comment (zhangkun83) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy]: About the rule-based model, I read the doc on Gradle website, but am still not completely clear.
IIUC, these rules are like tasks and are invoked based on their dependency relations, so I actually cannot find any explicit invocations of these methods. What is the first rule to be invoked, and by who?

Comment (Shad0w1nk) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy]: You got it mostly right. The rules are acting on the model. You can declare new element on the model with the @Model annotation. The @Default, @Mutate and @Finalize are simply to control when these run are applied on the model object. Each annotated rule method follow this simple convention: The first argument is the subject of the rule (what is been modified) and the following argument are the inputs of the rule. The argument for the @Model rule are all inputs (no subject as it's the return value). All input are guarantee to be completely configured before been used. Gradle take care of the dependencies between all the rule.

As for when, what and who of the rules, it's all handled by the internal of Gradle. When a plugin is applied, Gradle will look at the class and find the class extending the RuleSource. Then, it will scrap all the prototype of the annotated method and build the rules and it's dependencies. When Gradle need a specific element of the model, for example you want to execute a specific task, Gradle will call only the required rule (method) to configure that specific model.

This effectively make Gradle extremely lazy at the runtime level because it will only invoke what needs to be configure. This mainly the evaluation phase but does help with the execution phase. It prevent the old problem of "configuring the world" just to execute a simple task that is fairly independent.

The following videos really helped me to have a deep understanding of the model:
 - https://gradle.org/the-new-gradle-model/
 - https://gradle.org/rule-based-model-configuration-workshop-part-1/
 - https://gradle.org/rule-based-model-configuration-workshop-part-2/

I hope this help, don't hesitate if you have other question. I prefer you clearly understand.

Comment (zhangkun83) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy]: Thanks a lot. These videos are very helpful. Let me verify my understanding:

`TaskContainer` is always needed, which will trigger `configureGeneratedSourceSets()` and `finishConfiguringTheGenerateProtoTask()` since they have `TaskContainer` as their subjects. `configureGeneratedSourceSets()`, because of `ProtobufConfigurator` as its input, will trigger `protobuf()` that creates it.

`ModelMap<NativeComponentSpec>` is needed for compiling native project, which will trigger `createDefaultProtobufSourceSets()` then `bindDefaultProtoSourceSetToDefaultCppSourceSet()`.

Comment (Shad0w1nk) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy]: That sounds about right. Just some little note to make sure nothing was forgotten.

 * `finishConfiguringTheGenerateProtoTask()` will be called after `configureGeneratedSourceSets()` because of the mutation dependencies `@Finalize` vs `@Mutate`.
 * Before the rules `registerLanguageTransformation()` and `configureGeneratedSourceSets()`, the rule `resolveProtobufTools()` will be called as it take the `ProtobufConfigurator` as it's subject.
 * The `@LanguageType` annotation is a special one for doing enhancement of the core model. This one in particular register a new language type so we can create new language like:
```
sources {
  anotherSourceSet(ProtobufSourceSet) {...}
}
```
 * As for the rules that take `ModelMap<NativeComponentSpec>` as the subject, they will not necessarily configure the `ModelMap` completely as that type is model aware and will configure the item underneath it only when they are required. With the statement `components.all`, we are only stating that if anything in the `ModelMap<NativeComponentSpec>` is needed, apply the following action to it. If you have multiple component but you are only building a single one, the rules taking `ModelMap<NativeComponentSpec>` as the subject will be called but the action will only be called once as only one component is needed.

To have some additional insight on the rules and model elements, you can use the ./gradlew model which will display the various model elements and what rule is modifying them (and which plugin they are coming from). Very useful for seeing the bigger picture.

I hope this help a bit more, don't hesitate if you have more question.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    def gv = project.gradle.gradleVersion =~ "(\\d*)\\.(\\d*).*"
+    if (!gv || !gv.matches() || gv.group(1).toInteger() < 2 || gv.group(2).toInteger() < 7) {
+      println("You are using Gradle ${project.gradle.gradleVersion}: This version of the protobuf" +
+          " plugin for c++ requires minimum Gradle version 2.7")
+    }
+
+    // Provides the osdetector extension
+    project.pluginManager.apply('osdetector')
+
+    project.extensions.create("protobuf", ProtobufConfigurator, project, fileResolver);
```
Comment (zhangkun83) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy]: Since the user would use `model.protobuf`, why do you still need to create this extension?

Comment (Shad0w1nk) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy]: You are right that we could get away from creating the extension. However, the `ProtobufConfigurator` requires the instance of `project` which can only be reference from the `apply()` method. I didn't want to refactor `ProtobufConfigurator` in this PR. From what I could see from the Gradle code base, they usually create an extension in the `apply()` method then extract it from the `ExtensionContainer` in the model creation rule.

In other to get rid of this, we would have to refactor `ProtobufConfiguration` in such a way that it doesn't need a `Project` instance. I don't think this is trivial for now and would probably fit in the following PR which would try to bring closer the Java and C++ plugin. What do you think?

Comment (zhangkun83) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy]: Fair enough. Please add a TODO here to explain why it is needed at this moment and what the plan is.

Comment (Shad0w1nk) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy]: Done. Tell me if you need me to add anything else.

File: build.gradle [hunk 6]
```diff
@@ -61,16 +69,46 @@ repositories {
...
+
+  // We just want this path to be realize so we ignore the return type
+  vendorModelRegistry.realizeNode(ModelPath.path('components'))
+  def protoc = vendorModelRegistry.find(ModelPath.path('components.protoc.binaries.protocExecutable'), ModelType.of(NativeExecutableBinarySpec.class))
```
Comment (zhangkun83) [inline build.gradle]: Please limit your line length to 100, and follow [this guide](https://google.github.io/styleguide/javaguide.html#s4.4-column-limit) for line breaks.

File: vendor/build.gradle [hunk 7]
```diff
@@ -0,0 +1,168 @@
+project(':vendor') {
+  apply plugin: 'cpp'
+
+  model {
+    components {
+      // We are building protoc and protobuf from source for maximum compatibility
+      // across platform. Please note that we are using a very minimilistic
+      // port of the build to Gradle.
```
Comment (zhangkun83) [inline vendor/build.gradle]: Did you write these build configurations by hand? So, are you not using the build scripts that come with protobuf at all? (which is totally fine. I am just curious)

File: src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+          return "generate"
+        }
+
+        @Override
+        Class<? extends DefaultTask> getTaskType() {
+          return GenerateProtoTask
+        }
+
+        @Override
+        void configureTask(Task task, BinarySpec binary, LanguageSourceSet language, ServiceRegistry serviceRegistry) {
```
Comment (zhangkun83) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy]: To be accurate, "language" should be "sourceSet".

File: src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+          }
+        }
+      }
+    }
+  }
+
+  // TODO: This should probably be pushed into Gradle core.
+  public class CppFile implements TransformationFileType {}
+
+  private static class Proto implements LanguageTransform<ProtobufSourceSet, CppFile> {
```
Comment (zhangkun83) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy]: What is this class for?

File: src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+                      anotherCpp.lib it
+                    }
+                  }
+                }
+              }
+            }
+          }
+        """
+
+    and: "protobuf files across multiple source set that end up having the same include files"
```
Comment (zhangkun83) [inline src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy]: Do you mean "the same include file names?"

File: build.gradle [hunk 11]
```diff
@@ -61,16 +69,46 @@ repositories {
...
+  outputs.dir outputDir
+
+  doLast {
+    outputDir.mkdirs()
+    file("$outputDir/plugin-classpath.txt").text = sourceSets.main.runtimeClasspath.join("\n")
+  }
+}
+
+tasks.test.dependsOn tasks.createClasspathManifest
+if (findProject(':vendor') != null) {
```
Comment (zhangkun83) [inline build.gradle]: This if-block can be reduced to just one line:
```
tasks.test.dependsOn "vendor:protocExecutable",
    "vendor:protobufStaticLibrary", "vendor:protobufLiteStaticLibrary"
```

File: src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+
+    def pluginClasspath = pluginClasspathResource.readLines()
+        .collect { it.replace('\\', '\\\\') } // escape backslashes in Windows paths
+        .collect { "'$it'" }
+        .join(", ")
+
+    // Add the logic under test to the test build
+    buildFile << """
+          buildscript {
+            dependencies {
```
Comment (zhangkun83) [inline src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy]: Can you do the same trick for the other test subprojects, so that they no longer require "install" to test?

Comment (Shad0w1nk) [inline src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy]: We could look at making this possible as another PR. This was pull directly from the userguide: https://docs.gradle.org/2.7/userguide/test_kit.html#testKitFunctionalTestSpockClassesTestConfig. I don't see why we couldn't remove the "install" requirement.

File: src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy (hunk 10 above)
Comment (Shad0w1nk) [inline src/test/groovy/com/google/protobuf/gradle/ProtobufModelPluginTest.groovy]: Yes you are right. You have a better wording. I will update accordingly.

File: vendor/build.gradle (hunk 7 above)
Comment (Shad0w1nk) [inline vendor/build.gradle]: I extracted the information by hand from the Bazel build script. Since I didn't want to make the Bazel tools part of the environment (as it wasn't previously) I opted to include the build logic inside Gradle instead since it's perfectly capable to build the Protobuf library. In the future, we could extract the information on the fly which will make the script more flexible when changing the submodule tag.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy (hunk 9 above)
Comment (Shad0w1nk) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy]: This is the only class that needs to use internal interface when using the Gradle mechanize for automatically creating the appropriate language transform task for the declared source set. Basically, it describe that all source from ProtobufSourceSet will be transformed into CppFile according to the specification of this class.

I'm keeping an eye out for finding a better way (not using internal class/interface) to achieve this transformation. It was discuss in these threads:
 * https://groups.google.com/d/topic/gradle-dev/l4KTozshWXA/discussion
 * https://groups.google.com/d/topic/gradle-dev/jteaSNnnFtY/discussion

File: src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy (hunk 8 above)
Comment (Shad0w1nk) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy]: You are right, I will make the change.

File: build.gradle (hunk 11 above)
Comment (Shad0w1nk) [inline build.gradle]: Brilliant! I had a problem with the settings.gradle changing the behavior for the install task and I forgot about passing task dependencies as string instead of actual object reference. Way better way of doing this. I will make the change.

Comment (Shad0w1nk) [inline build.gradle]: It's not that trivial and trust me I would love if it could be. There is 2 problems:
 1- The install task act a bit differently because of the project include manipulation happening in settings.gradle. I have to check the availability of the `:vendor` project as during the install task, the project is not include which would crash the install process.
 2- The native use the model and the java use the legacy TaskContainer. This cause problem when trying to bridge task between both world. The only way I found was through the ModelRegistry. The design doc trying to solve this issue is https://github.com/gradle/gradle/blob/master/design-docs/task-bridging.md. Some discussion happen about this can be found here: https://groups.google.com/d/topic/gradle-dev/y9WmHyzCX-s/discussion

I'm hoping to find a better way. As soon as I find one this place will be the first to get this fix.

**Comment by Shad0w1nk** [comment]
@zhangkun83 The fixes have been pushed. The latest commit (message starting with "Merge branch 'native' of ...) seems to be an duplicate of a previous commit adding a TODO. Don't hesitate to ask for any additional fix.

File: build.gradle (hunk 11 above)
Comment (zhangkun83) [inline build.gradle]: I am confused. It works fine for me. What's the issue you encountered and how can I reproduce it?

**Comment by zhangkun83** [comment]
Please also apply the line limit to the other files (e.g., ProtobufModelPlugin.groovy)

**Comment by zhangkun83** [comment]
`ProtobufConfigurator.runTaskConfigClosures()` should be called before executing the tasks, in order to make the configurations in `protobuf.generateProtoTasks` take effect. Please cover this functionality in the tests.

File: build.gradle (hunk 11 above)
Comment (Shad0w1nk) [inline build.gradle]: I was also confused and I found the problem which I'm still debugging. On Mac OS X and probably Linux, your suggestion worked. However, on Windows, it doesn't seems to work. It fail with "Task with path 'vendor:protobufExecutable' not found in root project 'protobuf-gradle-plugin'". I will try to figure out why this happen as I prefer your solution by far!

Comment (zhangkun83) [inline build.gradle]: I think you should have `vendor:protocExecutable` instead :)

**Comment by Shad0w1nk** [comment]
Just to let you know, I'm still working on this patch. I'm working around the problem I got on Windows-only execution of Gradle. This may take a bit more time to work out.

**Comment by jaredsburrows** [comment]
@Shad0w1nk What is the ETA for the final PR updates?

**Comment by Shad0w1nk** [comment]
I got busy a bit with work. We ported to Gradle 2.13 and tried to complete the pull request. We are not there yet because of this problem: https://discuss.gradle.org/t/limitation-of-generatedby-in-native-software-model/17321

I'm hoping to try another work around sometime this month. I don't want to commit something that would break in the next release. I'm also keeping in mind the other language JVM, Python, etc. so we can reuse the work for anything related to the model. Given that what I previously did was breaking in newer release, this is why I'm been more cautious about the next pull request. I hope this is all good with you.

**Comment by nicknezis** [comment]
Is this ever going to be merged in? I would love for direct c++ support. Or is this pull request OBE?

**Comment by Shad0w1nk** [comment]
Good question @nicknezis. I haven't had much time to work on this. I still have good intentions of proposing a better PR for this work. However, lots have changed since I did this work. I now work on the native side of Gradle tool itself (through my other account @lacasseio). We are currently stabilizing the native support in Gradle as part of a [recent effort](https://github.com/gradle/gradle-native). Once this work is completed, I will revisit C++ support for this plugin as an example of the new effort. We just have to wait a bit longer. Given this, I will close this PR and open a new one once I get around to improving the current solution.

**Comment by nicknezis** [comment]
Thanks for the update. Let me know if I can help. If the native stuff is far along enough that I can at least try. (Not sure if there are other road blocks beyond just you having the cycles to work on it) 

I need this for a work project and am willing to contribute if possible. Would have to just solve my use case with an Exec task.
